sockeye/training.py

```diff
@@ -160,16 +168,68 @@ def sym_gen(seq_lens):
                     # target_embed: (batch, targ max len, embedding) -> (batch * targ max len, embedding)
                     target_embed = mx.sym.reshape(data=target_embed, shape=(-3, 0))
 
-                    context, attention = target_decoded_and_attention[1:]
+                    context, attention = target_decoded_and_context[1:]
 
                     # context: (batch_size * trg_seq_len, encoder_num_hidden)
                     context = mx.sym.reshape(data=context, shape=(-3, 0))
                     attention = mx.sym.reshape(data=attention, shape=(-3, 0))
 
+                if self.config.pointer_net_type != C.POINTER_NET_SUMMARY:
+                    # softmax_probs: (batch_size * target_seq_len, target_vocab_size+src_seq_len)
+                    softmax_probs = self.output_layer(target_decoded, attention=attention,
+                                                      context=context, target_embed=target_embed)
+                    loss_output = self.model_loss.get_loss(softmax_probs, labels)
+
+                else:
+                    context, attention, coverage, target_embed = target_decoded_and_context[1:]
+
+                    # target_decoded: (batch_size * target_seq_len, rnn_num_hidden)
+                    target_decoded = mx.sym.reshape(data=target_decoded, shape=(-3, 0))
+                    # context: (batch_size * trg_seq_len, encoder_num_hidden)
+                    context = mx.sym.reshape(data=context, shape=(-3, 0))
+                    # attention: (batch_size * target_seq_len, att_len)
+                    attention = mx.sym.reshape(data=attention, shape=(-3, 0))
+                    # coverage: (batch_size * target_seq_len, att_len)
+                    coverage = mx.sym.reshape(data=coverage, shape=(-3, 0))
+                    # target_embed: (batch_size * target_seq_len, target_embed_len)
+                    target_embed = mx.sym.reshape(data=target_embed, shape=(-3, -1))
+
+                    batch_size = self.batch_size * self.config.config_data.max_seq_len_target
+                    max_seq_len_source = self.config.config_data.max_seq_len_source
+                    max_seq_len_target = self.config.config_data.max_seq_len_target
+                    ext_vocab_size = self.config.config_embed_source.vocab_size + self.config.max_oov_words
+
+                    # prob_vocab: (batch_size * target_seq_len, vocab_size)
+                    # prob_source: (batch_size * target_seq_len, source_len)
+                    prob_vocab, prob_source = self.output_layer(target_decoded, context=context,
+                                                                          attention=attention,
+                                                                          target_embed=target_embed)
+
+                    # to correctly add source word probabilities based on the source word indices in the vocabulary
+                    # generate corresponding indices matrix
```

Comment:
I understand the net effect of the code below, but following the details is difficult. Can you add a comment above each line, describing what exactly that line is doing, and why? Basically, a detailed walk-through that will leave the reader understanding *how* the code sums up probabilities from the target vocabulary and from pointed words in the source sentence.

Note also that the [documentation for MXNet's `scatter_nd()` function](http://mxnet.apache.org/api/python/symbol/symbol.html#mxnet.symbol.scatter_nd) has a call-out claiming the behavior is undefined if there are duplicate indices, which can definitely occur. Do you have any idea of how to address this?